sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java

```diff
@@ -739,7 +739,6 @@ private void queueWork(int maximumMessageCount, Duration maxWaitTime,
         if (synchronousMessageSubscriber.compareAndSet(null, newSubscriber)) {
             asyncClient.receiveMessagesNoBackPressure().subscribeWith(newSubscriber);
         } else {
-            newSubscriber.dispose();
```

Comment:
If we dispose newSubscriber here, it will terminate the work and emitter. 
This is rarely case when multiple thread try to receiver message together.

Reply:
`synchronousMessageSubscriber` seems only allow one subscriber in the first place (that it only allow `null` to `newSubscriber`?

Seems it should throw exception for the 2nd thread here?

Reply:
On the other hand, if we keep 1st subscriber, why dispose the `newSubscriber` a problem? Anyway we put the `work` to 1st subscriber?

The problem appear to be that 2nd `newSubscriber` should not start the work (`addCap`?) before it is subscribed?

Reply:
>synchronousMessageSubscriber seems designed to only allow one subscriber in the first place

Right, we keep one subscriber `this.synchronousMessageSubscriber`, `newSubscriber` is to solve race condition in multiple threads. The expected behavior is clear `newSubscriber` when `this.synchronousMessageSubscriber` is not null, and add work into `this.synchronousMessageSubscriber`. But current issue is `newSubscriber.dispose()` will also terminate the work and complete the emitter, which is not what we want.

As we talked, maybe we could just use sync block here? or throw exception since it only happen when we start receiving message in mutli-thread together.

>how `requested` relate to `work`? If we has one `SynchronousMessageSubscriber` but multiple work, does this requested number for all work, or only current work?

`requested` is below to one `SynchronousMessageSubscriber` (we only maintain `this.SynchronousMessageSubscriber` to queue work). The `requested` represent how many message are requesting now, normally it represent for `CurrentWork` request number. So if it is 0, `CurrentWork` should be completed. 

There is one edge case: 1st work request receive 50 message (`requested` = 50), but it only receive 10 and work terminated by timeout(`requested` now change to 40). The 2nd work only receive 10 messages. After receive 10 messages, the 2nd work complete. `requested` change to 30. Then 3rd work will be updated by [SynchronousMessageSubscriber.java#L202](https://github.com/Azure/azure-sdk-for-java/blob/982b06d7163536c5aefc15be641ba52e9cbdf2fe/sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java#L202) to keep receiving message.

Current the code change is to take care when `requested` == 0, update current work to request more message.

Reply:
I would vote for sync block, since it is easier, and in normal case we would only use the sync block once for the whole ReceiverClient.

If current work is the only contributor to the `requested`, one possibility is to reset the `requested` if current work timeout/terminated?

Reply:
OK. `requested` seems fine as long as ctor is not called incorrectly (then disposed).